[Cookbook] Reading, writing data to key-value (shared preferences)

charafau opened the pull request:
After seeing list of [topics](https://github.com/flutter/flutter/issues/15297) for cookbook done by @brianegan  I thought about adding key-value section. Sorry for my english, I am open to suggestions. Thank you!

Hixie commented:
cc @brianegan for review

brianegan commented:
Hey, thanks for submitting this! Sorry for the delay, was out sick for a couple days. Reviewing.

brianegan commented on cookbook/persistance/key-value.md:
```diff
@@ -0,0 +1,102 @@
+permalink: /cookbook/persistence/key-value/
+---
+
+If you have a relatively small collection of key-values that you'd like to save, you should use the [shared_preferences](https://pub.dartlang.org/packages/shared_preferences) APIs
```
Maybe change to "You can use the shared_preferences plugin"?

brianegan commented on cookbook/persistance/key-value.md:
```diff
@@ -0,0 +1,102 @@
+
+If you have a relatively small collection of key-values that you'd like to save, you should use the [shared_preferences](https://pub.dartlang.org/packages/shared_preferences) APIs
+
+Normally we would have to write native platform integration for storing data for both platforms, but fortunetly [shared_preferences](https://pub.dartlang.org/packages/shared_preferences) plugin can be used for that. 
```
Typo, should be: "fortunately"

brianegan commented on cookbook/persistance/key-value.md:
```diff
@@ -0,0 +1,102 @@
+
+Normally we would have to write native platform integration for storing data for both platforms, but fortunetly [shared_preferences](https://pub.dartlang.org/packages/shared_preferences) plugin can be used for that. 
+
+Shared preferences plugin wraps `NSUserDefaults` (on iOS) and `SharedPreferences` (on Android), providing a persistent store for simple data. Data is persisted to disk automatically and asynchronously.
```
"The shared_preferences plugin wraps `NSUserDefaults` on iOS and `SharedPreferences` on Android" -- no need for the extra parens `()` 

Hrm, I find the line "Data is persisted to disk automatically" a bit confusing. Don't know if that sentence is necessary! We're handing the plugin some data manually, so  "automatic" sounds a bit funny.

Overall, I'd consider combining this part with the previous sentence to form a complete paragraph.

brianegan commented on cookbook/persistance/key-value.md:
````diff
@@ -0,0 +1,102 @@
+  shared_preferences: "<newest version>"
+```
+
+## Saving, reading and removing data
````
Maybe we could create 3 sections here::

  1. Saving Data
  2. Reading Data
  3. Removing Data

In the example below, the order is Read (null), Set, Delete, which might be a bit funny for an introduction.

brianegan commented on cookbook/persistance/key-value.md:
```diff
@@ -0,0 +1,102 @@
+
+In example above we load data from `counter` key, if it not exists, `0` is returned. 
+
+For saving data we call `set` method. Note that data is persisted on disk automatically and asynchronously. If you want to get notified when data is saved use `commit()` function. 
```
Maybe "persisted asynchronously" rather than "automatically"? The word "automatically" seems to imply it's doing something magical for us.

brianegan commented on cookbook/persistance/key-value.md:
```diff
@@ -0,0 +1,102 @@
+
+For saving data we call `set` method. Note that data is persisted on disk automatically and asynchronously. If you want to get notified when data is saved use `commit()` function. 
+
+Adequately setter and getter methods are available for all primitive types.
```
"Adequately setter and getter methods are available for all primitive types." -- Please remove "Adequately" at the beginning of this sentence.

brianegan commented on cookbook/persistance/key-value.md:
````diff
@@ -0,0 +1,102 @@
+}
+```
+
+## Testing support
````
Love it!

brianegan commented on cookbook/persistance/key-value.md:
```diff
@@ -0,0 +1,102 @@
+  ));
+}
+
+_incrementCounter() async {
```
Rather than printing to the console, could we please display the counter on screen? It could basically be the same as the Flutter Starter Template + a `FutureBuilder` to fetch the initial value.

charafau commented:
Thanks for review, hope I will be about to fix that today.

brianegan commented:
@charafau Thanks much! Whenever you have time :)

brianegan commented on cookbook/persistance/key-value.md:
````diff
@@ -0,0 +1,102 @@
+You can populate SharedPreferences with initial values in your tests by running this code:
+
+```dart
+const MethodChannel('plugins.flutter.io/shared_preferences')
````
Is this valid code? Should the `const` be here?

charafau commented:
@brianegan  I applied your suggestions and fixed the typos. As for example, I added that new one (modified version of sample project) but instead of using `FutureBuilder` I used async method and called it from `initState()`  method. If you want me to change it I will use `FutureBuilder` then. I thought that `FutureBuilder` is a little bit overkill for that.

brianegan commented:
@charafau Excellent, thanks for taking the time to go through the changes! I'll review once more :)

brianegan commented on cookbook/persistance/key-value.md:
```diff
@@ -0,0 +1,175 @@
+
+If you have a relatively small collection of key-values that you'd like to save, you can use the [shared_preferences](https://pub.dartlang.org/packages/shared_preferences) plugin.
+
+Normally we would have to write native platform integration for storing data for both platforms, but fortunately [shared_preferences](https://pub.dartlang.org/packages/shared_preferences) plugin can be used for that. Shared preferences plugin wraps `NSUserDefaults` on iOS and `SharedPreferences` on Android, providing a persistent store for simple data.
```
"... for both platforms. Fortunately, the shared_prefs..." 

Two sentences: "but fortunately" is a bit awkward;  put "the" before shared_preferences

brianegan commented on cookbook/persistance/key-value.md:
```diff
@@ -0,0 +1,175 @@
+
+## Saving data
+
+In order to access store simply use `SharedPreferences` class.
```
"In order to access the store, we can use..."

In general, I'm trying to avoid the words "just" or "simply" -- I've gotten feedback from newcomers that even if something seems simple, it might be difficult to understand!

brianegan commented on cookbook/persistance/key-value.md:
```diff
@@ -0,0 +1,175 @@
+permalink: /cookbook/persistence/key-value/
+---
+
+If you have a relatively small collection of key-values that you'd like to save, you can use the [shared_preferences](https://pub.dartlang.org/packages/shared_preferences) plugin.
```
The text in this recipe switches between "you" and "we". Could you please use "we" throughout? I'm not sure it's the best choice, but I've gone with "we" in other recipes, and good to be consistent.

charafau commented:
Thanks for reviewing that. I fixed all `you` to `we` and `your` to `our` 👍  (and other changes requested)

brianegan commented on cookbook/persistance/key-value.md:
```diff
@@ -0,0 +1,175 @@
+
+In example above we load data from `counter` key, if it not exists, `0` is returned. 
+
+For saving data we call `set` method. Note that data is persisted asynchronously. If we want to get notified when data is saved use `commit()` function. 
```
Does this content belong in the "Saving Data" section? I think you could combine it with the "To persist key-value data, we can use..." line

brianegan commented on cookbook/persistance/key-value.md:
```diff
@@ -0,0 +1,175 @@
+
+## Testing support
+
+We can populate SharedPreferences with initial values in our tests by running this code:
```
For folks who are new, could you please describe where to place this code? e.g. in a `setupAll` method or the test file?

charafau replied:
I put simple note about this. I think in future, when test page will be written, we can change that to link to testing.

brianegan commented on cookbook/persistance/key-value.md:
```diff
@@ -0,0 +1,175 @@
+
+## Supported types
+
+While is easy and convenient to use key-value storage, it has limitations:
```
"While *it* is easy"

brianegan commented on cookbook/persistance/key-value.md:
```diff
@@ -0,0 +1,175 @@
+  - `int`, `double`, `bool`, `string` and `string list`
+- It's not designed to store a lot of data so it's not suitable as application cache.
+
+If we want to read more about Shared preferences on Android please visit [Shared preferences documentation](https://developer.android.com/guide/topics/data/data-storage.html#pref) on Android developers website.
```
"For more information about Shared Preferences on Android, please visit..."

brianegan commented on cookbook/persistance/key-value.md:
````diff
@@ -0,0 +1,169 @@
+int counter = (prefs.getInt('counter') ?? 0) + 1;
+```
+
+In example above we load data from the `counter` key, if it not exists, `0` is returned. 
````
"In this example we'll load data from the counter key. If it *does not* exist, 0 is returned."

brianegan commented on cookbook/persistance/key-value.md:
```diff
@@ -0,0 +1,169 @@
+
+## Saving data
+
+To persist key-value data, we can use `SharedPreferences` class.
```
"we can use *the* SharedPreferences

brianegan commented on cookbook/persistance/key-value.md:
```diff
@@ -0,0 +1,169 @@
+## Saving data
+
+To persist key-value data, we can use `SharedPreferences` class.
+For saving data we call `set` method. Note that data is persisted asynchronously. If we want to get notified when data is saved use `commit()` function. 
```
"we call *the* set method"

"If we want to be notified when data is saved, use *the* `commit()` function. "

brianegan commented on cookbook/persistance/key-value.md:
````diff
@@ -0,0 +1,169 @@
+  });
+```
+
+Code above should be placed inside test file under `test` folder.
````
*The* code above.

Haha, I know adding *the* everywhere seems crazy, but it's idiomatic English :)

charafau replied:
I really appreciate you fixing my bad English, so please continue with my next request :) hope you won't be mad at me haha

brianegan replied:
Haha, never gone be mad when folks are contributing. Thanks again!

brianegan commented:
Cool, looking good overall! @filiph or @mjohnsullivan, Could you please have a read-through and provide feedback / merge it up if it looks ok?

mjohnsullivan commented:
LGTM
